File: android/app/src/main/java/com/sakredbody/app/AppearancePlugin.java
package com.sakredbody.app;

import android.app.Activity;
import android.content.SharedPreferences;
import android.graphics.Color;
import android.graphics.drawable.ColorDrawable;
import android.view.View;
import android.view.Window;

import androidx.core.view.WindowInsetsControllerCompat;

import com.getcapacitor.JSObject;
import com.getcapacitor.Plugin;
import com.getcapacitor.PluginCall;
import com.getcapacitor.PluginMethod;
import com.getcapacitor.annotation.CapacitorPlugin;

/**
 * The parts of the screen the WebView does not draw.
 *
 * ── What @capacitor/status-bar leaves undone ──────────────────────────────
 *
 * That plugin sets `setAppearanceLightStatusBars` and stops. On a phone with
 * gesture navigation there is a second bar at the bottom, and from Android 15
 * it is transparent by platform decision — so the pill is drawn straight onto
 * whatever the web layer painted, in whatever contrast the system was last
 * told to use. In Dark that is light-on-dark and correct by accident. The
 * moment a member picks Light it is a light pill on limestone: still there,
 * still swipeable, and invisible.
 *
 * ── Why this is an app-local plugin ───────────────────────────────────────
 *
 * It could have been a package under plugins/ next to health-sync. It is a
 * dozen lines of platform call with no npm surface, no iOS/Android shared
 * contract worth versioning and nothing another app would install, so it
 * lives in the app module and is registered by hand in MainActivity.
 *
 * ── Why the window background is written down ─────────────────────────────
 *
 * A cold launch has no WebView yet. `styles.xml` names one windowBackground
 * for every member, and it has to be ink because the splash is ink — but the
 * frames *after* the splash tears down and before the first paint arrives use
 * the same value, and for a member in Light that is a dark flash at the end of
 * a launch rather than the beginning. So the resolved ground is stored here
 * and MainActivity reads it back on the next launch, which is the only place
 * the choice can be known before the web layer exists to say it.
 */
@CapacitorPlugin(name = "SakredAppearance")
public class AppearancePlugin extends Plugin {

    static final String PREFS = "sakred.appearance";
    static final String KEY_DARK = "dark";
    static final String KEY_INK = "ink";

    @PluginMethod
    public void apply(PluginCall call) {
        final String theme = call.getString("theme", "dark");
        final boolean dark = !"light".equals(theme);
        final String ink = call.getString("ink");

        final Activity activity = getActivity();
        if (activity == null) {
            call.reject("No activity.");
            return;
        }

        // Parsed on this side of the bridge, so a malformed colour is a
        // rejected call rather than a crash on the UI thread.
        int ground = 0;
        boolean haveGround = false;
        if (ink != null && !ink.isEmpty()) {
            try {
                ground = Color.parseColor(ink);
                haveGround = true;
            } catch (IllegalArgumentException e) {
                call.reject("Not a colour: " + ink);
                return;
            }
        }

        final int groundColor = ground;
        final boolean paintGround = haveGround;

        activity.runOnUiThread(() -> {
            Window window = activity.getWindow();
            View decor = window.getDecorView();

            WindowInsetsControllerCompat controller =
                new WindowInsetsControllerCompat(window, decor);
            // "Light bars" means light *background*, therefore dark icons —
            // the same inversion @capacitor/status-bar's Style names carry,
            // and the same trap. Light theme → light bars → dark pill.
            controller.setAppearanceLightNavigationBars(!dark);
            // Set here as well as by the status-bar plugin. The two calls
            // agree, and a member who has never opened the health screen has
            // never caused that plugin to load.
            controller.setAppearanceLightStatusBars(!dark);

            if (paintGround) {
                // Behind the WebView, so a keyboard opening or the recents
                // card resuming shows this rather than the launch ink.
                window.setBackgroundDrawable(new ColorDrawable(groundColor));
            }
        });

        SharedPreferences.Editor edit =
            activity.getSharedPreferences(PREFS, Activity.MODE_PRIVATE).edit();
        edit.putBoolean(KEY_DARK, dark);
        if (haveGround) edit.putString(KEY_INK, ink);
        edit.apply();

        JSObject result = new JSObject();
        result.put("applied", true);
        result.put("dark", dark);
        call.resolve(result);
    }
}

File: ios/App/App/AppearancePlugin.swift
import Capacitor
import UIKit

/**
 The parts of the screen UIKit draws, not the web layer.

 ── What the status bar plugin leaves undone ──────────────────────────────

 `@capacitor/status-bar` sets the status bar's contrast and nothing else. On
 iOS the appearance a member picks also has to reach the pieces UIKit renders
 on the app's behalf: the keyboard, the text-selection callout, the scroll
 indicators, the sheet chrome, and the ground revealed by a rubber-band
 overscroll. All of those follow `overrideUserInterfaceStyle` on the window,
 and with no `UIUserInterfaceStyle` in Info.plist they follow the *system*
 instead — so a member who chose Light on a phone in night mode types into a
 black keyboard, and one who chose Dark on a bright phone gets a white one.

 ── Why the choice is written down ────────────────────────────────────────

 The window exists before the WebView has loaded, so at launch there is a
 stretch where UIKit has to be told something and only the last run knows
 what. SceneDelegate reads this back; here it is kept current.

 ── Why not just set UIUserInterfaceStyle in Info.plist ───────────────────

 That is a build-time constant. It would pin every member to one appearance,
 which is the problem rather than the fix.
 */
@objc(SakredAppearance)
public class AppearancePlugin: CAPPlugin, CAPBridgedPlugin {
    public let identifier = "SakredAppearance"
    public let jsName = "SakredAppearance"
    public let pluginMethods: [CAPPluginMethod] = [
        CAPPluginMethod(name: "apply", returnType: CAPPluginReturnPromise)
    ]

    static let defaultsKey = "sakred.appearance.dark"

    /// What the window should be in, given what was last applied.
    static func storedStyle() -> UIUserInterfaceStyle {
        // `object(forKey:)` rather than `bool(forKey:)`: the latter answers
        // false for a key that was never written, which would put a first
        // launch in Light. Absent means "not chosen yet", and the app is dark
        // until told otherwise.
        guard let stored = UserDefaults.standard.object(forKey: defaultsKey) as? Bool else {
            return .dark
        }
        return stored ? .dark : .light
    }

    @objc func apply(_ call: CAPPluginCall) {
        let dark = call.getString("theme", "dark") != "light"

        UserDefaults.standard.set(dark, forKey: AppearancePlugin.defaultsKey)

        DispatchQueue.main.async {
            // Every window, not just the plugin's own: a keyboard or an
            // action sheet can be presented in a different one, and a single
            // window left on the system style is the one the member notices.
            for scene in UIApplication.shared.connectedScenes {
                guard let windowScene = scene as? UIWindowScene else { continue }
                for window in windowScene.windows {
                    window.overrideUserInterfaceStyle = dark ? .dark : .light
                }
            }
            call.resolve(["applied": true, "dark": dark])
        }
    }
}
